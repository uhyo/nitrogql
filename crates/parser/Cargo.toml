[package]
name = "nitrogql-parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
nitrogql-ast = { path = "../ast" }
nitrogql-error = { path = "../error" }
pest = "2.5.5"
pest_derive = "2.5.5"
thiserror = "2.0.12"

[dev-dependencies]
insta = "1.28.0"
nitrogql-printer = { path = "../printer" }
sourcemap-writer = { path = "../sourcemap-writer" }
