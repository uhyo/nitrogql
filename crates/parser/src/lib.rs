mod parser;
mod tests;

pub use parser::{ParseError, parse_operation_document, parse_type_system_document};
