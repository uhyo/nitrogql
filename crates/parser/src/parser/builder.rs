//! This module builds ast from raw parser result.

use self::{
    operation::build_executable_definition, type_system::build_type_system_definition_or_extension,
    utils::PairExt,
};

use super::Rule;
use nitrogql_ast::{
    operation_ext::OperationDocumentExt, type_system::TypeSystemOrExtensionDocument,
};
use pest::iterators::Pairs;

mod base;
mod directives;
mod operation;
mod selection_set;
mod r#type;
mod type_system;
mod utils;
mod value;

pub fn build_operation_document(pairs: Pairs<Rule>) -> OperationDocumentExt {
    if let Some(pair) = pairs.into_iter().next() {
        match pair.as_rule() {
            Rule::ExecutableDocument => {
                let position = pair.to_pos();
                let definitions: Vec<_> = pair
                    .into_inner()
                    .filter(|pair| pair.is_rule(Rule::ExecutableDefinition))
                    .map(build_executable_definition)
                    .collect();
                return OperationDocumentExt {
                    position,
                    definitions,
                };
            }
            rule => panic!("Unexpected Rule {:?}", rule),
        }
    }
    panic!("Empty document")
}

pub fn build_type_system_or_extension_document(
    pairs: Pairs<Rule>,
) -> TypeSystemOrExtensionDocument {
    if let Some(pair) = pairs.into_iter().next() {
        match pair.as_rule() {
            Rule::TypeSystemExtensionDocument => {
                let definitions: Vec<_> = pair
                    .into_inner()
                    .filter(|pair| pair.is_rule(Rule::TypeSystemDefinitionOrExtension))
                    .map(build_type_system_definition_or_extension)
                    .collect();
                return TypeSystemOrExtensionDocument { definitions };
            }
            rule => panic!("Unexpected Rule {:?}", rule),
        }
    }
    panic!("Empty document")
}
