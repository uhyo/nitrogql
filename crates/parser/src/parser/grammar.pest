// This grammar implements the GraphQL spec,
// plus the following extension:
// - #import syntax


// B.3 Lexical Tokens

WHITESPACE = _{ "\u{FEFF}" | "\u{0009}" | " " | NEWLINE | "," }
COMMENT = _{ "#" ~ !(&ext_KEYWORD_import ~ ext_ImportStatementContent) ~ CommentCharacter* ~ NEWLINE }
CommentCharacter = { !NEWLINE ~ ANY }

Token = { Punctuator | Name | IntValue | FloatValue | StringValue }
Punctuator = { "!" | "$" | "&" | "(" | ")" | "..." | ":" | "=" | "@" | "[" | "]" | "{" | "|" | "}" }
Name = @{ NameStart ~ NameContinue* }
NameStart = @{ ASCII_ALPHA | "_" }
NameContinue = @{ ASCII_ALPHANUMERIC | "_" }

IntValue = @{ IntegerPart ~ !("." | NameStart) }
IntegerPart = @{ "-"? ~ ( "0" | (ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*)) }

FloatValue = @{
  (IntegerPart ~ FractionalPart ~ ExponentPart ~ !("." | NameStart) ) |
  (IntegerPart ~ FractionalPart ~ !("." | NameStart) ) |
  (IntegerPart ~ ExponentPart ~ !("." | NameStart) )
}
FractionalPart = @{ "." ~ ASCII_DIGIT* }
ExponentPart = @{ ^"e" ~ ("+" | "-")? ~ ASCII_DIGIT+ }

StringValue = ${ EmptyStringValue | NormalStringValue | BlockStringValue }
EmptyStringValue = @{ "\"\"" ~ !("\"") }
BlockStringValue = @{ "\"\"\"" ~ BlockStringCharacter* ~ "\"\"\"" }
NormalStringValue = ${ "\"" ~ StringCharacter+ ~ "\"" }
StringCharacter = ${
  EscapedUnicodeBrace |
  EscapedUnicode4 |
  EscapedCharacter | 
  NormalStringCharacter
}
BlockStringCharacter = @{
  !"\"\"\"" ~
  ("\\\"\"\"" | ANY)
}

EscapedUnicodeBrace = ${ "\\u" ~ "{" ~ EscapedUnicodeBraceDigits ~ "}" }
EscapedUnicodeBraceDigits = @{ ASCII_HEX_DIGIT+ }
EscapedUnicode4 = @{ "\\u" ~ ASCII_HEX_DIGIT{4} }
EscapedCharacter = @{ "\\" ~ ("\"" | "\\" | "/"	| "b" | "f" | "n" | "r" | "t") }
NormalStringCharacter = @{ !("\"" | "\\" | NEWLINE ) ~ ANY }

// keywords
KEYWORD_query = @{ "query" ~ !NameContinue }
KEYWORD_mutation = @{ "mutation" ~ !NameContinue }
KEYWORD_subscription = @{ "subscription" ~ !NameContinue }
KEYWORD_fragment = @{ "fragment" ~ !NameContinue }
KEYWORD_on = @{ "on" ~ !NameContinue }
KEYWORD_true = @{ "true" ~ !NameContinue }
KEYWORD_false = @{ "false" ~ !NameContinue }
KEYWORD_null = @{ "null" ~ !NameContinue }
KEYWORD_extend = @{ "extend" ~ !NameContinue }
KEYWORD_schema = @{ "schema" ~ !NameContinue }
KEYWORD_scalar = @{ "scalar" ~ !NameContinue }
KEYWORD_type = @{ "type" ~ !NameContinue }
KEYWORD_implements = @{ "implements" ~ !NameContinue }
KEYWORD_interface = @{ "interface" ~ !NameContinue }
KEYWORD_union = @{ "union" ~ !NameContinue }
KEYWORD_enum = @{ "enum" ~ !NameContinue }
KEYWORD_input = @{ "input" ~ !NameContinue }
KEYWORD_directive = @{ "directive" ~ !NameContinue }
KEYWORD_repeatable = @{ "repeatable" ~ !NameContinue }
ext_KEYWORD_import = @{ "import" ~ !NameContinue }
ext_KEYWORD_from = @{ "from" ~ !NameContinue }

// punctuations
PUNC_equal = { "=" }
PUNC_open_brace = { "{" }
PUNC_close_brace = { "}" }

// B.4 Document Syntax

// Document = { Definition+ }
// Definition = { ExecutableDefinition | TypeSystemDefinitionOrExtension }

ExecutableDocument = { SOI ~ ExecutableDefinition+ ~ EOI }

ExecutableDefinition = { OperationDefinition | FragmentDefinition | ext_ImportStatement }
OperationDefinition = {
  OperationType ~ Name? ~ VariablesDefinition? ~ Directives? ~ SelectionSet | SelectionSet
}

OperationType = { KEYWORD_query | KEYWORD_mutation| KEYWORD_subscription }
SelectionSet = { "{" ~ Selection+ ~ "}"}

Selection = { Field | FragmentSpread | InlineFragment }

Field = { Alias? ~ Name ~ Arguments? ~ Directives? ~ SelectionSet? }
Alias = { Name ~ ":" }
Arguments = { "(" ~ Argument+ ~ ")" }
Argument = { Name ~ ":" ~ Value }

FragmentSpread = { "..." ~ FragmentName ~ Directives? }
InlineFragment = { "..." ~ TypeCondition? ~ Directives? ~ SelectionSet }

FragmentDefinition = { KEYWORD_fragment ~ FragmentName ~ TypeCondition ~ Directives? ~ SelectionSet }

FragmentName = { !KEYWORD_on ~ Name }
TypeCondition = { KEYWORD_on ~ NamedType }

Value = {
  Variable |
  IntValue |
  FloatValue |
  StringValue |
  BooleanValue |
  NullValue |
  EnumValue |
  ListValue |
  ObjectValue
}

BooleanValue = { KEYWORD_true | KEYWORD_false }
NullValue = { KEYWORD_null }
EnumValue = { !(KEYWORD_true | KEYWORD_false | KEYWORD_null) ~ Name }

ListValue = {
  "[" ~ "]" |
  "[" ~ Value+ ~ "]"
}
ObjectValue = {
  "{" ~ "}" |
  "{" ~ ObjectField+ ~ "}"
}
ObjectField = { Name ~ ":" ~ Value }

VariablesDefinition = { "(" ~ VariableDefinition+ ~ ")" }
VariableDefinition = { Variable ~ ":" ~ Type ~ DefaultValue? ~ Directives? }

Variable = { "$" ~ Name }

DefaultValue = { "=" ~ Value }

Type = { NonNullType | NamedType | ListType }
NamedType = { Name }
ListType = { "[" ~ Type ~ "]" }
NonNullType = { (NamedType ~ "!") | (ListType ~ "!") }

Directives = { Directive+ }
Directive = { "@" ~ Name ~ Arguments? }


TypeSystemExtensionDocument = { SOI ~ TypeSystemDefinitionOrExtension+ ~ EOI }

TypeSystemDefinitionOrExtension = { TypeSystemDefinition | TypeSystemExtension }
TypeSystemDefinition = { SchemaDefinition | TypeDefinition | DirectiveDefinition }
TypeSystemExtension = { SchemaExtension | TypeExtension }

SchemaDefinition = { Description? ~ KEYWORD_schema ~ Directives? ~ RootOperationTypeDefinitions }
SchemaExtension = {
  (KEYWORD_extend ~ KEYWORD_schema ~ Directives? ~ RootOperationTypeDefinitions) |
  (KEYWORD_extend ~ KEYWORD_schema ~ Directives ~ !"{")
}

RootOperationTypeDefinitions = { "{" ~ RootOperationTypeDefinition+ ~ "}" }
RootOperationTypeDefinition = { OperationType ~ ":" ~ NamedType }
Description = { StringValue }

TypeDefinition = {
  ScalarTypeDefinition |
  ObjectTypeDefinition |
  InterfaceTypeDefinition |
  UnionTypeDefinition |
  EnumTypeDefinition |
  InputObjectTypeDefinition
}

TypeExtension = {
  ScalarTypeExtension |
  ObjectTypeExtension |
  InterfaceTypeExtension |
  UnionTypeExtension |
  EnumTypeExtension |
  InputObjectTypeExtension
}

ScalarTypeDefinition = { Description? ~ KEYWORD_scalar ~ Name ~ Directives? }
ScalarTypeExtension = { KEYWORD_extend ~ KEYWORD_scalar ~ Name ~ Directives? }

ObjectTypeDefinition = {
  Description? ~ KEYWORD_type ~ Name ~ ImplementsInterfaces? ~ Directives? ~ FieldsDefinition |
  Description? ~ KEYWORD_type ~ Name ~ ImplementsInterfaces? ~ Directives ~ !("{")
}
ObjectTypeExtension = {
  KEYWORD_extend ~ KEYWORD_type ~ Name ~ ImplementsInterfaces? ~ Directives? ~ FieldsDefinition |
  KEYWORD_extend ~ KEYWORD_type ~ Name ~ ImplementsInterfaces? ~ Directives ~ !("{") |
  KEYWORD_extend ~ KEYWORD_type ~ Name ~ ImplementsInterfaces ~ !("{")
}

ImplementsInterfaces = { KEYWORD_implements ~ "&"? ~ NamedType ~ ("&" ~ NamedType)* }

FieldsDefinition = { "{" ~ FieldDefinition+ ~ "}" }
FieldDefinition = { Description? ~ Name ~ ArgumentsDefinition? ~ ":" ~ Type ~ Directives? }

ArgumentsDefinition = { "(" ~ InputValueDefinition+ ~ ")" }

InputValueDefinition = { Description? ~ Name ~ ":" ~ Type ~ DefaultValue? ~ Directives? }

InterfaceTypeDefinition = {
  Description? ~ KEYWORD_interface ~ Name ~ ImplementsInterfaces? ~ Directives? ~ FieldsDefinition |
  Description? ~ KEYWORD_interface ~ Name ~ ImplementsInterfaces? ~ Directives? ~ !"{"
}
InterfaceTypeExtension = {
  KEYWORD_extend ~ KEYWORD_interface ~ Name ~ ImplementsInterfaces? ~ Directives? ~ FieldsDefinition |
  KEYWORD_extend ~ KEYWORD_interface ~ Name ~ ImplementsInterfaces? ~ Directives? ~ !"{" |
  KEYWORD_extend ~ KEYWORD_interface ~ Name ~ ImplementsInterfaces ~ !"{"
}

UnionTypeDefinition = { Description? ~ KEYWORD_union ~ Name ~ Directives? ~ "=" ~ UnionMemberTypes? }
UnionMemberTypes = { "|"? ~ NamedType ~ ("|" ~ NamedType)* }
UnionTypeExtension = {
  KEYWORD_extend ~ KEYWORD_union ~ Name ~ Directives? ~ "=" ~ UnionMemberTypes |
  KEYWORD_extend ~ KEYWORD_union ~ Name ~ Directives
}

EnumTypeDefinition = {
  Description? ~ KEYWORD_enum ~ Name ~ Directives? ~ EnumValuesDefinition |
  Description? ~ KEYWORD_enum ~ Name ~ Directives? ~ !"{"
}
EnumValuesDefinition = { "{" ~ EnumValueDefinition+ ~ "}" }
EnumValueDefinition = { Description? ~ EnumValue ~ Directives? }
EnumTypeExtension = {
  KEYWORD_extend ~ KEYWORD_enum ~ Name ~ Directives? ~ EnumValuesDefinition |
  KEYWORD_extend ~ KEYWORD_enum ~ Name ~ Directives? ~ !"{"
}

InputObjectTypeDefinition = {
  Description? ~ KEYWORD_input ~ Name ~ Directives? ~ InputFieldsDefinition |
  Description? ~ KEYWORD_input ~ Name ~ Directives? ~ !"{"
}
InputFieldsDefinition = { "{" ~ InputValueDefinition+ ~ "}" }
InputObjectTypeExtension = {
  KEYWORD_extend ~ KEYWORD_input ~ Name ~ Directives? ~ InputFieldsDefinition |
  KEYWORD_extend ~ KEYWORD_input ~ Name ~ Directives? ~ !"{"
}

DirectiveDefinition = {
  Description? ~ KEYWORD_directive ~ "@" ~ Name ~ ArgumentsDefinition? ~ KEYWORD_repeatable? ~ KEYWORD_on ~ DirectiveLocations
}
DirectiveLocations = { "|"? ~ DirectiveLocation ~ ("|" ~ DirectiveLocation)* }
DirectiveLocation = { ExecutableDirectiveLocation | TypeSystemDirectiveLocation }
ExecutableDirectiveLocation = @{
  ("QUERY" | "MUTATION" | "SUBSCRIPTION" | "FIELD" | "FRAGMENT_DEFINITION" |
  "FRAGMENT_SPREAD" | "INLINE_FRAGMENT" | "VARIABLE_DEFINITION"
  ) ~ !NameContinue
}
TypeSystemDirectiveLocation = @{
  ("SCHEMA" | "SCALAR" | "OBJECT" | "FIELD_DEFINITION" | "ARGUMENT_DEFINITION" |
  "INTERFACE" | "UNION" | "ENUM_VALUE" | "ENUM"  | "INPUT_OBJECT" | "INPUT_FIELD_DEFINITION"
  ) ~ !NameContinue
}

ext_ImportStatement = { "#" ~ ext_ImportStatementContent }
ext_ImportStatementContent = _{ ext_KEYWORD_import ~ ext_NameOrAsterisk+ ~ ext_KEYWORD_from ~ StringValue }
ext_NameOrAsterisk = { Name | "*" }