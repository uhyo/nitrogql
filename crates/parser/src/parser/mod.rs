use nitrogql_ast::{TypeSystemOrExtensionDocument, base::Pos, operation_ext::OperationDocumentExt};
use nitrogql_error::PositionedError;
use pest::Parser;
use pest_derive::Parser;
use thiserror::Error;

use self::builder::{build_operation_document, build_type_system_or_extension_document};

mod builder;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RawParser;

#[derive(Error, Debug)]
#[error("Parse error: {0}")]
pub struct ParseErrorMessage(String);

/// Struct that expresses parse error.
#[derive(Debug)]
pub struct ParseError {
    position: Pos,
    message: String,
}

impl ParseError {
    /// Extracts error message.
    pub fn into_message(self) -> String {
        self.message
    }
}

impl From<pest::error::Error<Rule>> for ParseError {
    fn from(error: pest::error::Error<Rule>) -> Self {
        // convert 1-based line and column to 0-based
        let (line, col) = match error.line_col {
            pest::error::LineColLocation::Pos((line, column)) => (line - 1, column - 1),
            pest::error::LineColLocation::Span((line, column), _) => (line - 1, column - 1),
        };
        let position = Pos::new(line, col);
        let message = error.variant.message().into_owned();

        ParseError { position, message }
    }
}

impl From<ParseError> for PositionedError {
    fn from(value: ParseError) -> Self {
        let position = value.position;
        let additional_info = vec![];
        let inner = ParseErrorMessage(value.message).into();

        PositionedError::new(inner, Some(position), additional_info)
    }
}

pub fn parse_operation_document(document: &str) -> Result<OperationDocumentExt, ParseError> {
    let res = RawParser::parse(Rule::ExecutableDocument, document)?;

    Ok(build_operation_document(res))
}

pub fn parse_type_system_document(
    document: &str,
) -> Result<TypeSystemOrExtensionDocument, ParseError> {
    let res = RawParser::parse(Rule::TypeSystemExtensionDocument, document)?;

    Ok(build_type_system_or_extension_document(res))
}
