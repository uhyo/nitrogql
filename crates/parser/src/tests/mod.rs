#[cfg(test)]
mod operation {
    use insta::assert_snapshot;

    use crate::parser::parse_operation_document;
    use nitrogql_printer::GraphQLPrinter;
    use sourcemap_writer::JustWriter;

    #[test]
    fn simple_query() {
        assert_snapshot!(print_graphql(
            parse_operation_document("query { foo bar }").unwrap()
        ))
    }
    #[test]
    fn query_with_name() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "
                query sample {
                    foo
                    bar {
                        nested1
                        nested2
                    }
                 }
"
            )
            .unwrap()
        ))
    }
    #[test]
    fn query_with_variables() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "
                query sample($foo: Int!, $bar: Int!) {
                    foo @dedede(foo: $foo)
                 }
"
            )
            .unwrap()
        ))
    }

    #[test]
    fn query_with_directives() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "
                query sample($foo: [Int!]!) @a @b(c: D) {
                    foo
                    bar
                 }
"
            )
            .unwrap()
        ))
    }

    #[test]
    fn mutation_with_fragment() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "
                mutation($foo: Boolean) {
                    foo
                    ...Fragment
                }
                "
            )
            .unwrap()
        ))
    }
    #[test]
    fn subscription_with_inline_fragment() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "
                subscription a {
                    foo
                    ... on F {
                        bar
                        baz
                    }
                }
                "
            )
            .unwrap()
        ));
    }
    #[test]
    fn comments() {
        assert_snapshot!(print_graphql(
            parse_operation_document(
                "# Comment
query #comment
sample#comment
($foo: Int!, #comment
# # # # #
     $bar: Int!) {
    # comment
    foo
    # comment
    bar
} # comment
"
            )
            .unwrap()
        ));
    }

    fn print_graphql<T: GraphQLPrinter>(value: T) -> String {
        let mut result = String::new();
        let mut writer = JustWriter::new(&mut result);
        value.print_graphql(&mut writer);
        result
    }
}

#[cfg(test)]
mod definition {
    use crate::parser::parse_type_system_document;
    use insta::assert_snapshot;
    use nitrogql_printer::GraphQLPrinter;
    use sourcemap_writer::JustWriter;

    #[test]
    fn scalar_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                scalar Int
                \"Description\"
                scalar String @string

                extend scalar Int @heyhey(foo: \"bar\")
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn object_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                type Foo {
                    foo: String!
                    bar: String!
                }
                type Bar implements I @wow {
                    \"this is foo\" foo: String @wow 
                }
                \"\"\"
                Description of type
                \"\"\"
                type Baz implements I & J {
                    func(arg1: Int): Int
                }

                extend type Foo {
                    foo2: String!
                }
                extend type Bar implements J & K & L
                extend type Baz @wow
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn interface_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                interface MyI {
                    foo: String!
                    bar: [Int!]!
                }
                \"Gaooo ü¶Å\" interface _Lion implements MyI {
                    foobar: [[[Int]]]
                }
                interface aiu implements MyI & _Lion @wow

                extend interface MyI {
                    hey: Hey!
                }
                extend interface _Lion implements MyI2 @abc
                extend interface aiu @heyhey
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn union_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                union ABC = A | B | C,
                \"XYZ Dragon Cannon\"union XYZ @x @y @z = | X | Y
                | Z

                extend union ABC = D
                extend union XYZ @xyz
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn enum_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                enum Ehh { E H h }
                \"This\\nis\\nenum\" enum EEE {
                    E @desc(message: \"Hello\")
                    E2 @desc(message: null)
                    E3 @desc(message: false)
                }

                extend enum Ehh { EEEEE, EEEE, \"heyhey\" EEE }
                extend enum EEE @eeeee
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn input_object_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                input Foo {
                    foo: String!
                    bar: String!
                }
                input Bar @wow {
                    \"this is foo\" foo: String = \"aaa\"  @wow
                    bar: Bar
                }
                \"\"\"
                Description of type
                \"\"\"
                input Baz {
                    baz: Int! = 3
                }

                extend input Foo {
                    foo2: String! @foo(num: 2)
                }
                extend input Bar @barber
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn directive_definition() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "
                directive @foo on QUERY | MUTATION | SUBSCRIPTION
                \"\"\"
                Hey \"\"hey\"\" \\\"\"\"Hey\\\"\"\"
                \"\"\"
                directive @bar repeatable on INPUT_FIELD_DEFINITION
                directive @baz(arg1: Int! @arg, arg2: Int! @arg) on INPUT_OBJECT
                "
            )
            .unwrap()
        ));
    }

    #[test]
    fn comments() {
        assert_snapshot!(print_graphql(
            parse_type_system_document(
                "# Comment
scalar Date # Comment
# Comment
# comment # comment ### comment

type #comment
Foo # comment
# comment
{
    # comment
    foo: String! # comment
    # comment
    bar: String! # comment
}
"
            )
            .unwrap()
        ));
    }

    fn print_graphql<T: GraphQLPrinter>(value: T) -> String {
        let mut result = String::new();
        let mut writer = JustWriter::new(&mut result);
        value.print_graphql(&mut writer);
        result
    }
}
