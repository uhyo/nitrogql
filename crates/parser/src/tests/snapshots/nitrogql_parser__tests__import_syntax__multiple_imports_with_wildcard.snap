---
source: crates/parser/src/tests/mod.rs
expression: "print_graphql(parse_operation_document(r#\"\n                #import Frag1, * from \"./frag1.graphql\"\n                query Foo {\n                    foo {\n                        ...Frag1\n                        ...Frag2\n                    }\n                }\n                \"#).unwrap())"
---
#import Frag1, * from "./frag1.graphql"
query Foo {
  foo {
    ... Frag1
    ... Frag2
  }
}
